### DISCUSSION TITLE
Residency guarantee for per-room sockets on distributed architectures

### DISCUSSION THREAD

**Discussion by nicholaswmin** [socketio/socket.io#4990] [Q&A] [answered]
Assuming a 10 dyno/process formation, using `ioredis` and the adapters on socketio v4, ioredis v5.3

In the following example I'm instantiating something and assigning it to `socket.data`, like so:

```js
io.use((socket, next) => {
  socket.data = new Cow()
 
  next()
});

io.on('connection', () => {
  socket.join(id_session)

  socket.data.moo() // moooo.
})    
```
Should I expect that this instance will survive for the duration of the connection or is there any chance of some serialising/deserialising taking place which is going to cause it to lose it's type and methods?

I'm only asking about the duration of the connection, i.e if the client disconnects the state is of course lost.

**Comment by darrachequesne** [comment, marked as answer]
Hi! Yes, the `socket` object is guaranteed to stay on the same Socket.IO server for the duration of the connection.

However, please note that if you enable [Connection state recovery](https://socket.io/docs/v4/connection-state-recovery), then in that case the `socket` object will indeed be serialised upon disconnection and recreated upon reconnection if the recovery is successful.

**Reply by nicholaswmin** [reply]
Hey Damien - another quick question - do all the participants of a room stay on the same server ?

**Reply by darrachequesne** [reply]
@nicholaswmin no, the participants of a given room are not guaranteed to be on the same server.
